src/callstack.cr

```diff
@@ -0,0 +1,190 @@
+      bt = data.as(typeof(callstack))
+      ip = Pointer(Void).new(LibUnwind.get_ip(context))
+      bt << ip
+
+      {% if flag?(:gnu) && flag?(:i686) %}
+        # This is a workaround for glibc bug: https://sourceware.org/bugzilla/show_bug.cgi?id=18635
+        # The unwind info is corrupted when `makecontext` is used.
+        # Stop the backtrace here. There is nothing interest beyond this point anyway.
+        if CallStack.makecontext_range.includes?(ip)
+          return LibUnwind::ReasonCode::END_OF_STACK
+        end
+      {% end %}
+
+      LibUnwind::ReasonCode::NO_REASON
+    end
+
+    LibUnwind.backtrace(backtrace_fn, callstack.as(Void*))
+    callstack
+  end
+
+  struct RepeatedFrame
+    getter ip : Void*, count : Int32
+
+    def initialize(@ip : Void*)
+      @count = 0
+    end
+
+    def incr
+      @count += 1
+    end
+  end
+
+  def self.print_backtrace
+    backtrace_fn = ->(context : LibUnwind::Context, data : Void*) do
+      last_frame = data.as(RepeatedFrame*)
+      ip = Pointer(Void).new(LibUnwind.get_ip(context))
+      if last_frame.value.ip == ip
+        last_frame.value.incr
+      else
+        print_frame(last_frame.value) unless last_frame.value.ip.address == 0
+        last_frame.value = RepeatedFrame.new ip
+      end
+      LibUnwind::ReasonCode::NO_REASON
+    end
+
+    rf = RepeatedFrame.new(Pointer(Void).null)
+    LibUnwind.backtrace(backtrace_fn, pointerof(rf).as(Void*))
+    print_frame(rf)
+  end
+
+  private def self.print_frame(repeated_frame)
+    frame = decode_frame(repeated_frame.ip)
+    if frame
+      offset, sname = frame
+      if repeated_frame.count == 0
+        LibC.printf "[%ld] %s +%ld\n", repeated_frame.ip, sname, offset
+      else
+        LibC.printf "[%ld] %s +%ld (%ld times)\n", repeated_frame.ip, sname, offset, repeated_frame.count + 1
+      end
+    else
+      if repeated_frame.count == 0
+        LibC.printf "[%ld] ???\n", repeated_frame.ip
+      else
+        LibC.printf "[%ld] ??? (%ld times)\n", repeated_frame.ip, repeated_frame.count + 1
+      end
+    end
+  end
+
+  private def decode_backtrace
+    @callstack.compact_map do |ip|
+      file, line, column = CallStack.decode_line_number(ip)
+      if file == "??"
+        file_line_column = "??"
+      else
+        next if @@skip.includes?(file)
+        file_line_column = "#{file} #{line}:#{column}"
+      end
+
+      if frame = CallStack.decode_frame(ip)
+        _, sname = frame
+        function = String.new(sname)
+      else
+        function = "???"
+      end
+
+      "0x#{ip.address.to_s(16)}: #{function} at #{file_line_column}"
+    end
+  end
+
+  {% if flag?(:freebsd) || flag?(:linux) %}
+    @@dwarf_line_numbers : Debug::DWARF::LineNumbers?
+
+    protected def self.dwarf_line_numbers?
+      @@dwarf_line_numbers ||= Debug::ELF.open(PROGRAM_NAME) do |elf|
+        elf.read_section?(".debug_line") do |sh, io|
+          Debug::DWARF::LineNumbers.new(io, sh.size)
+        end
+      end
```

Comment:
I chose to memoize the decompressed DWARF line numbers section in memory. This consumes some memory (around 8MB for the Crystal compiler) but this is only generated when printing the backtraces, and it will save time to avoid decompressing the whole section each time an exception is raised, for a service that rescues exceptions and logs them.